server/src/test/java/org/elasticsearch/common/util/concurrent/PrioritizedThrottledTaskRunnerTests.java

```diff
@@ -69,119 +73,120 @@ public void onFailure(Exception e) {
-            assertThat(taskRunner.runningTasks(), equalTo(0));
-        });
+        assertTrue(executedCountDown.await(10, TimeUnit.SECONDS));
         assertThat(taskRunner.queueSize(), equalTo(0));
+        assertNoRunningTasks(taskRunner);
     }
 
     public void testTasksRunInOrder() throws Exception {
-        final int n = randomIntBetween(1, threadPool.info(ThreadPool.Names.GENERIC).getMax());
-        final int enqueued = randomIntBetween(2 * n, 10 * n);
+        final int n = randomIntBetween(1, maxThreads);
         // To check that tasks are run in the order (based on their priority), limit max running tasks to 1
         // and wait until all tasks are enqueued.
-        CountDownLatch workerBlocker = new CountDownLatch(1);
-        PrioritizedThrottledTaskRunner<TestTask> taskRunner = new PrioritizedThrottledTaskRunner<>("test", 1, executor) {
-            @Override
-            protected void pollAndSpawn() {
-                try {
-                    workerBlocker.await();
-                } catch (InterruptedException e) {
-                    throw new RuntimeException(e);
-                }
-                super.pollAndSpawn();
+        final var barrier = new CyclicBarrier(2);
+        final var taskRunner = new PrioritizedThrottledTaskRunner<TestTask>("test", 1, executor);
+
+        taskRunner.enqueueTask(new TestTask(() -> {
+            try {
+                barrier.await(10, TimeUnit.SECONDS);
+                barrier.await(10, TimeUnit.SECONDS);
+            } catch (Exception e) {
+                throw new AssertionError("unexpected", e);
             }
-        };
+        }, getRandomPriority()));
+
+        barrier.await(10, TimeUnit.SECONDS);
+
+        final int enqueued = randomIntBetween(2 * n, 10 * n);
         List<Integer> taskPriorities = new ArrayList<>(enqueued);
         List<Integer> executedPriorities = new ArrayList<>(enqueued);
+        final var enqueuedBarrier = new CyclicBarrier(enqueued + 1);
+        final var executedCountDown = new CountDownLatch(enqueued);
         for (int i = 0; i < enqueued; i++) {
             final int priority = getRandomPriority();
             taskPriorities.add(priority);
-            new Thread(() -> taskRunner.enqueueTask(new TestTask(() -> executedPriorities.add(priority), priority))).start();
+            new Thread(() -> {
+                awaitBarrier(enqueuedBarrier);
+                taskRunner.enqueueTask(new TestTask(() -> {
+                    executedPriorities.add(priority);
+                    executedCountDown.countDown();
+                }, priority));
+                awaitBarrier(enqueuedBarrier);
+            }).start();
         }
-        assertBusy(() -> assertThat(taskRunner.queueSize(), equalTo(enqueued)));
-        assertThat(taskRunner.runningTasks(), equalTo(0));
-        workerBlocker.countDown();
+        awaitBarrier(enqueuedBarrier);
+        awaitBarrier(enqueuedBarrier);
```

Comment:
The distinction between the two awaits here are not very clear to understand. One is to unblock the enqueued tasks at L108? Could you add a comment here please.

Reply:
Maybe a more concrete question would be: why are we sync'ing all the threads? Wouldn't all the tasks enqueued in the threads be blocked anyway?

Reply:
Really this is just so the `enqueueTask()` calls are as concurrent as possible. Without this, we're relying on going round the thread-creation loop quickly enough to generate concurrent activity. I expect we'd catch issues eventually either way, but empirically this seems to make failures a little easier to reproduce.

Reply:
I added clarifying comments in 00ef643f835, hope that's ok.